jupyter_server_proxy/websockify.py

```diff
@@ -0,0 +1,89 @@
+"""
+A simple translation layer between tornado websockets and asyncio stream
+connections.
+
+This subsumes the functionality of websockify
+(https://github.com/novnc/websockify) without needing an extra proxy hop
+or process through with all messages pass for translation.
+"""
+
+import asyncio
+
+from .handlers import NamedLocalProxyHandler, SuperviseAndProxyHandler
+
+class WebsockifyProtocol(asyncio.Protocol):
+    """
+    A protocol handler for the proxied stream connection.
+    Sends any received blocks directly as websocket messages.
+    """
+    def __init__(self, handler):
+        self.handler = handler
+
+    def data_received(self, data):
+        "Send the buffer as a websocket message."
+        self.handler._record_activity()
+        # async "semi-synchronous", waiting is only needed for control flow and errors:
+        self.handler.write_message(data, binary=True)
+
+    def connection_lost(self, exc):
+        "Close the websocket connection."
+        self.handler.log.info(f"Websockify {self.handler.name} connection lost: {exc}")
+        self.handler.close()
+
+class WebsockifyHandler(NamedLocalProxyHandler):
+    """
+    HTTP handler that proxies websocket connections into a backend stream.
+    All other HTTP requests return 405.
+    """
+    def _create_ws_connection(self, proto: asyncio.BaseProtocol):
+        "Create the appropriate backend asyncio connection"
+        loop = asyncio.get_running_loop()
+        if self.unix_socket is not None:
+            self.log.info(f"Websockify {self.name} connecting to {self.unix_socket}")
+            return loop.create_unix_connection(proto, self.unix_socket)
+        else:
+            self.log.info(f"Websockify {self.name} connecting to port {self.port}")
+            return loop.create_connection(proto, 'localhost', self.port)
+
+    async def proxy(self, port, path):
+        raise web.HTTPError(405, "websockets only")
+
+    async def proxy_open(self, host, port, proxied_path=""):
+        """
+        Open the backend connection. host and port are ignored (as they are in
+        the parent for unix sockets) since they are always passed known values.
+        """
+        transp, proto = await self._create_ws_connection(lambda: WebsockifyProtocol(self))
+        self.ws_transp = transp
+        self.ws_proto = proto
+        self._record_activity()
+        self.log.info(f"Websockify {self.name} connected")
+
+    def on_message(self, message):
+        "Send websocket messages as stream writes, encoding if necessary."
+        self._record_activity()
+        if hasattr(self, "ws_transp"):
```

Comment:
Is there a reason to add this guard here explicitly, given that `on_message` would not be called before `proxy_open`?

Reply:
I think that's true as it stands, though on reviewing I realize `proxy_open` now should probably do a `_check_host_allowlist` and maybe other checks that may result in different situations. I'll look at it again with the latest code and clean up if possible.

Reply:
Or wait, no that's not relevant here since we're always proxying to a fixed place. (Sorry, haven't looked at this for a bit.)